Ldexp out of range

cleonard530 opened the pull request:
Fixes #153069

Updated ldexp to calculate $a\cdot2^b$ as $2^{\log_2(a) + b}$. This helps fix the issue where $2^b$ is not in range for a specific data type (such as $2^{-25}$ or $2^{20}$ with torch.half) but $a\cdot2^b$ is (such as $2\cdot2^{-25}$ or $2^{-10}\cdot 2^{20}$). 

The issue was happening because $2^b$ was being calculated first, which would go out of range and mess up the rest of the calculation. Using $2^{\log_2(a) + b}$, we calculate ${\log_2(a) + b}$ first, thus the exponential will only go out of range if the whole expression is out of range. 

This also fixes the issue #133265, that was closed since it was a duplicate of the other.

@ngimel, @isuruf 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

cleonard530 committed:
updated ldexp to use log identites calulate the function. This can help handle out of bounds error (particularly  dtype==Half). Still ned to fix for complex numbers.

cleonard530 committed:
Set it so that complex and bool values fall back to old implementation.

cleonard530 committed:
Added unit test to check for fix. Removed unused includes from BinaryOps.cpp.

cleonard530 commented:
@ngimel, I first implemented the attempt in #154211, but because it uses std::ldexp, only integer values were allowed for the exponential argument (other). This would break backwards compatibility and would have required a bunch of changes to the current unit test to avoid passing in floats and complex values.

cleonard530 committed:
Updated ldexp to convert reduced floating types to float while doing calulations and then convert back when calculations are done.

cleonard530 committed:
removed #includes that I added for previous commits but are unnecessary now.

cleonard530 commented:
> @cleonard530, that would change the return type when reduced floating types are given right?
> 
> How about we revive #154211 and fallback to the existing inaccurate method when floating point exponents are given?

Hi @isuruf. No, it still provides the reduces return type. This is accomplished because of the update I made to the ldexp method that sets the return type and calls ldexp_out instead of calling at::mul. This is tested by test_ldexp in `pytorch/test/test_binary_ufuncs.py` where all the test passed (running `python -m pytest test/test_binary_ufuncs.py -k ldexp -v `) . 

Also,  I ran this code to time the method before and after my update:
```
n = 1000000000
exp_t = -25*torch.ones(n, dtype=torch.int32)
tensor2 = (-2*torch.ones(n)).to(torch.float16)
time_function(torch.ldexp, True, tensor2, exp_t)
```
Before my updates it took about 0.8 seconds and after it only took about 0.5 seconds (both CPU and CUDA). I think this is because before it was converting to float32 to calculate $2^b$ then back to float16 then back to float32 to calculate $a\cdot2^b$ then back to float16 again, whereas now it doesn't need to do that intermediate conversion from float32 to float16.

cleonard530 commented:
> @cleonard530, that would change the return type when reduced floating types are given right?
> 
> How about we revive #154211 and fallback to the existing inaccurate method when floating point exponents are given?

@isuruf, Also, by the existing inaccurate method, do you mean the method that uses $2^{\log_2(a) + b}$ or the method that is currently on the main branch?

isuruf commented:
> the method that is currently on the main branch?

this one

ngimel commented:
Support for non-integer exps is accidental, tbh I'm not against removing it

cleonard530 committed:
Updated ldexp to use std::ldexp on cpu and ::ldexp on CUDA when then input types for 'input' and 'output' are float and int respectively.

ngimel commented:
@cleonard530 we can keep old implementation for non-int exponent, no harm in that. Thank you!

ngimel commented:
Sorry CI was out for a long time, let's try rebasing.

ngimel commented on aten/src/ATen/native/cuda/BinaryMiscOpsKernels.cu:
```diff
@@ -69,11 +69,20 @@ void xlog1py_kernel_cuda(TensorIteratorBase& iter) {
 
+void ldexp_kernel_cuda(TensorIteratorBase& iter) {
+  AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.input_dtype(0), "ldexp_cuda", [&] {
+    gpu_kernel_with_scalars(iter, []GPU_LAMBDA(scalar_t x, int exp) -> scalar_t {
```
btw I don't think you need `gpu_kernel_with_scalars` here, because ldexp doesn't support scalar overloads, just `gpu_kernel` should be enough

ngimel commented:
This apparently is breaking differentiability support

cleonard530 committed:
Updated the ldexp function to not use the _out variant since mul_out does not allow automatic differentiation.  Also removed unnecessary ";" that was causing build issues.

cleonard530 committed:
Updated gpu_kernel_with_scalars to gpu_kernel since ldexp doesn't support scalar overloads.

cleonard530 committed:
updated empty() to empty_like() since it was set to be like 'self' anyways.

cleonard530 committed:
Made the helper function _ldexp_int_exponent a static inline function and added the pointwise tag to ldexp.Tensor in native_functions.yaml. These should fix the current CI failures.

cleonard530 committed:
Added CompositeExplicitAutograd to ldexp since it uses the fast kernels now. Defined the gradients in derivatives.yaml and updated generation and testing tools to support the new Autograd setup.

cleonard530 commented:
@ngimel, yes you are right. The differentiation needs to be defined explicitly now. I have added  CompositeExplicitAutograd to native_functions.yaml and defined the gradient in derivatives.yaml file. I also made updates to fix the previous failed CI test. Thanks!

ngimel commented:
@pytorchbot rebase -b main

cleonard530 committed:
Added is_meta() check to ldexp so that when the tensor is meta it uses the preexisting meta checks from at::mul and at::mul_out. Also, updated the derivatives.yaml file to include .conj() for complex gradients.

ngimel commented:
@cleonard530 test failures are related?

ngimel commented:
Looks like `.conj()` is missing somewhere in the grad formula?

cleonard530 committed:
Needed to register decompositions for ldexp since it was no longer CompositeImplicitAutograd and changed to CompositeExplicitAutograd. This also fixed the meta_tensor issue I had with one of the test. Also needed to update the pointwise_ops ldexp entry since I added the pointwise tag to ldexp.Tensor.

cleonard530 commented:
@ngimel, sorry I missed that. 
I added the .conj() to the _derivatives.yaml_. 
I also added a **register_decomposition** for **ldexp** in _torch/_decomp/decompositions.py_ since it uses CompositeExplicitAutograd now, and updated the  **pointwise_ops** list in _torch/distributed/tensor/_ops/_pointwise_ops.py_ since I added the pointwise tag to **ldexp** in _native_functiosn.yaml_

cleonard530 committed:
Updated register_decomposition for ldexp to handle dtyps correctly.  Ran "EXPECTTEST_ACCEPT=1 python test/test_decomp.py HasDecompTest.test_aten_core_operators" to add ldexp and its varients to the list in test/expect/HasDecompTest.test_aten_core_operators.expect

cleonard530 committed:
Forgot to handle the register_decomposition case when both self and other are Integer (including bool) types. Also needed to change tensor construction from torch.tensor to self.new_full since fake CUDA tensors on a CPU using FakeTensorMode doesn't work with torch.tensor.

ngimel commented:
Separate issue is fine, I don't want to block this PR on that

ngimel commented:
@cleonard530 inductor tests are failing. You can either add ldexp decomp to inductor decompositions and later redo proper codegen that would allow inductor to generate `ldexp` directly, not causing overflow, or do it in this PR

cleonard530 committed:
Added aten.ldexp to the inductor_decompositions list. removed ldexp for the TORCH_LIBRARY_IMPL(aten, FuncTorchBatchedDecomposition, m) list since it is no longer an implicit autograd and added it to the TORCH_LIBRARY_IMPL(aten, FuncTorchBatched, m) list to work with vmap. Also, I had to increase the tolerance on test_vmap_autograd_grad slightly for the est_vmap_autograd_grad_ldexp_cuda_float32
test to pass.

cleonard530 committed:
Added ldexp.Tensor to the CHECK_STRIDES_SKIPS in test.meta.py. This seems to be a known issue and the CHECK_STRIDE_SKIPS list is how similar binary operations bypass this issue.

cleonard530 committed:
Added a print statement that I forgot to delete.

ngimel commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -14293,15 +14293,6 @@ def sample_inputs_alias_copy(op_info, device, dtype, requires_grad, **kwargs):
-                        # automatic differentiation, but one of the arguments requires grad
-                        # https://github.com/pytorch/pytorch/issues/68966
-                        DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_variant_consistency_eager'),
-                        DecorateInfo(unittest.expectedFailure, 'TestMathBits', 'test_conj_view'),
```
Probably makes sense to bring these skips back until proper inductor generation is fixed

cleonard530 committed:
Added the ldexp test skips back into pytorch/torch/testing/_internal/common_methods_invocations.py but changed the unittest.expectedFailure to unittest.skip

cleonard530 committed:
Added a comment explaining that we set unittest.skip since the inductor tests have errors even though the eager test pass.

cleonard530 commented:
@pytorchbot merge

cleonard530 commented:
@ngimel, can you approve this so that it can be merged?

ngimel commented:
@pytorchbot merge
